### DISCUSSION TITLE
Is there a way to make certain color ranges transparent?

### DISCUSSION THREAD

**Discussion by formula4** [ImageMagick/ImageMagick#4606] [General] [answered]
Hello all!

I am working with satellite data in JPG format, which two examples are : 

![image](https://user-images.githubusercontent.com/56749434/146634242-2601fe0d-bdcf-439a-9089-4816a0d7e04e.png)

and 

![image](https://user-images.githubusercontent.com/56749434/146634397-052ddc2f-076e-4e00-9257-04ac8017b185.png)

I only need the bright colors (lightning data - the neon green, yellow, reds). The rest of the image is background. The imagery might be day or night. 

Since I have a simple range of colors that I need (and all others can be discounted) ... is it possible to make every pixel that is not within the desired color range transparent? (in TIF format)? This would essentially be a JPG > TIF conversion.

I know how to convert to TIF using magick, and I know how to make specific colors transparent in magick. But I don't know if what I am wanting to do is possible.

Edit : To be clear, I do not need the clouds, water or any of the background whatsoever. I am just trying to isolate the overlay data only.

**Comment by fmw42** [comment, marked as answer]
What version of ImageMagick and what platform/OS?  If working with IM 7, then you can use -color-threshold to threshold on a range of colors, so that your desired range becomes white and the rest black. Save that as a mask and put it into the alpha channel of the image so that your colors remain and your background becomes transparent.

Input:

![146634397-052ddc2f-076e-4e00-9257-04ac8017b185](https://user-images.githubusercontent.com/13458218/146648150-33cd3329-b0b7-4e50-8789-be68b3ecaf49.png)


Unix syntax:

```
magick 146634397-052ddc2f-076e-4e00-9257-04ac8017b185.png \
\( +clone -color-threshold "srgb(20,175,0)-srgb(255,255,75)" \) \
-alpha off -compose copy_opacity -composite \
result.tif
```

Windows syntax:

```
magick 146634397-052ddc2f-076e-4e00-9257-04ac8017b185.png ^
( +clone -color-threshold "srgb(20,175,0)-srgb(255,255,75)" ) ^
-alpha off -compose copy_opacity -composite ^
result.tif
```

[result.tif.zip](https://github.com/ImageMagick/ImageMagick/files/7739560/result.tif.zip)

**Reply by formula4** [reply]
Thank for your your assistance, I didn't know this would be possible! This is why I love IM. I am using ImageMagick 7.1.0-17 Q16-HDRI x64 **2021-12-04** on Windows 10.

The range of colors I need to isolate goes into the purple/magenta hues and I have had trouble adjusting the script to accommodate them without removing the rest of the image.

![image](https://user-images.githubusercontent.com/56749434/146649093-649ecf63-cb9b-432b-b969-ca43d8e52ed6.png)

Your script above works fine except that theres some gaps in the data :  (different part of the image but same concept)

![image](https://user-images.githubusercontent.com/56749434/146649148-3e13824b-a9ac-4ba5-b8db-9deb1863eaa8.png)

Is this still easy to fix using the -color-threshold function? My idea otherwise is to manually add in allowed colors (since there isn't a large range, I could likely do something like : 

srgb(20,175,0),srgb(20,176,1),srgb(20,177,2)... etc etc. There may be many manual inputs I would have to make, but it would guarantee all of the data is saved. Any ideas here? 

For your amusement, I edited your script to be "srgb(20,175,0)-srgb(255,8,19)" (greens to purple) but this made the entire image white.

**Reply by fmw42** [reply]
I just put an example. You may need to split ranges. So you have to do the processing twice or more. Here I do the green/yellow and then the orange/red/pink. I save as PNG so that I can view the result here, but you can save as TIF

```
magick 146649093-649ecf63-cb9b-432b-b969-ca43d8e52ed6.png -alpha off \
\( -clone 0 -color-threshold "srgb(20,175,0)-srgb(255,255,95)" \) \
\( -clone 0 -color-threshold "srgb(200,0,0)-srgb(255,200,255)" \) \
\( -clone 1,2 -compose plus -composite \) \
-delete 1,2 \
-alpha off -compose copy_opacity -composite \
result.png
```
![result](https://user-images.githubusercontent.com/13458218/146652021-c9e5e0d1-9c74-4049-8ea3-494f3d284fde.png)

**Reply by formula4** [reply]
This is working great! Now that I know how to split ranges, I can do this with other data. Thank you for your assistance.

**Reply by formula4** [reply]
I noticed some other shades of magenta, so I decided to add another range. However it is giving me some trouble.

This is another set of data I observed :
![image](https://user-images.githubusercontent.com/56749434/146655802-897d2e06-6e89-4412-84e0-371f7f15ee36.png)


 Here is a modified script :

magick image.jpg -alpha off ( -clone 0 -color-threshold srgb(20,175,0)-srgb(255,255,95) ) ( -clone 0 -color-threshold srgb(200,0,0)-srgb(255,200,255) ) ( -clone 0 -color-threshold srgb(170,2,252)-srgb(186,71,252) ) ( -clone 1,2 -compose plus -composite ) -delete 1,2 -alpha off -compose copy_opacity -composite result.png

Note my 3rd range in there. Of course this does not work by itself. How else do I edit this for it to process the 3rd range? I tried "-clone 1,2,3" thinking that may be it, but it did not work. After some digging in the documentation and toying with the script I am not sure what to do. This is very close!

**Reply by fmw42** [reply]
assuming your range is OK, try

`magick image.jpg -alpha off ( -clone 0 -color-threshold srgb(20,175,0)-srgb(255,255,95) ) ( -clone 0 -color-threshold srgb(200,0,0)-srgb(255,200,255) ) ( -clone 0 -color-threshold srgb(170,2,252)-srgb(186,71,252) ) ( -clone 1-3 -clamp -evaluate-sequence add ) -delete 1-3 -alpha off -compose copy_opacity -composite result.png`

Or

magick image.jpg -alpha off ( -clone 0 -color-threshold srgb(20,175,0)-srgb(255,255,95) ) ( -clone 0 -color-threshold srgb(200,0,0)-srgb(255,200,255) ) ( -clone 0 -color-threshold srgb(170,2,252)-srgb(186,71,252) ) ( -clone 1-3 -clamp -poly "1,1 1,1 1,1"  ) -delete 1-3 -alpha off -compose copy_opacity -composite result.png

If not, post your input so I can see what actual range you need to add.  But snibgo's solution may be the simplest if you have do this for multiple regions.

**Reply by formula4** [reply]
This still is leaving some blank areas. My input is : 

![image](https://user-images.githubusercontent.com/56749434/146656156-96812fe0-981f-4f0d-ba84-dd076369c917.png)

I think part of the problem is that I do not know the exact valid range for the data. With these purples and magentas - it is very close to the maximum. I keep coming across small parts where it is missing still. With that said, this is not my area of expertise although I am trying. As for snibgo's solution, I am still trying to get that to work. It is not removing anywhere near the range that has to be removed so I cam still playing around with it.

**Comment by snibgo** [comment]
Another possibility is to observe that the required colours have high chroma, the "C" channel of HCL. Windows CMD syntax:
```
magick weather.png ( +clone -colorspace HCL -channel 1 -separate
 +channel -threshold 50% ) -alpha off -compose CopyOpacity -composite out.png
```
This also leaves opaque part of the logo at bottom-left.

**Reply by formula4** [reply]
I noticed this. Thank you for the observation and solution. It appears there were two ways to isolate the data, by color range and high chroma. I had no idea it would be this easy, I was prepared for battle. Long live IM.

**Comment by fmw42** [comment]
```
magick 146656156-96812fe0-981f-4f0d-ba84-dd076369c917.png -alpha off \
\( -clone 0 -color-threshold "srgb(0,115,0)-srgb(255,255,95)" \) \
\( -clone 0 -color-threshold "srgb(200,0,0)-srgb(255,200,255)" \) \
\( -clone 0 -color-threshold "srgb(150,35,225)-srgb(210,95,255)" \) \
\( -clone 1-3 -clamp -evaluate-sequence add \) \
-delete 1-3 \
-alpha off -compose copy_opacity -composite \
result.png

```
![result](https://user-images.githubusercontent.com/13458218/146656535-6c549dec-16f8-45d5-9680-855b56b65d4f.png)

You won't get the dark green at the bottom of the rectangle in the top left corner, because it is too desaturated of color and is too close gray of parts of your background. So if you got that to show, you would also get some of the background image to start showing as well.

**Reply by formula4** [reply]
So far this is working great! I will keep playing this, but this might be picking up nearly all the colors now.

**Comment by snibgo** [comment]
> As for snibgo's solution, I am still trying to get that to work. It is not removing anywhere near the range that has to be removed so I cam still playing around with it.

It works fine for me. As I said, it is Windows CMD format. If you are using Windows BAT, you need to double the percentage sign.

**Reply by formula4** [reply]
It's working great now! I need to dive deeper in command line stuff. Can't thank you enough.

**Reply by formula4** [reply]
Just a last question here, at the end of the script I must convert to TIF.  I tried :  magick convert out.png latest.tif ...
![image](https://user-images.githubusercontent.com/56749434/146657908-3742a26f-7fe1-4ef2-8975-f3a64172b725.png)

This resulted in the background coming back. Do PNGs have layers? I did not think this was the case.  From what I see online, this should have converted it. I feel like a troublemaker asking again, but I was sure this command would have converted.

**Reply by fmw42** [reply]
PNGs do not have layers. But my example does not make layers. So your convert should work fine, except use "magick" not "magick convert"

**Reply by fmw42** [reply]
Just change the suffix on my output from .png to .tif and it should work fine

```
magick 146656156-96812fe0-981f-4f0d-ba84-dd076369c917.png -alpha off ^
( -clone 0 -color-threshold "srgb(0,115,0)-srgb(255,255,95)" ) ^
( -clone 0 -color-threshold "srgb(200,0,0)-srgb(255,200,255)" ) ^
( -clone 0 -color-threshold "srgb(150,35,225)-srgb(210,95,255)" ) ^
( -clone 1-3 -clamp -evaluate-sequence add \) ^
-delete 1-3 ^
-alpha off -compose copy_opacity -composite ^
result.tif
```

**Reply by formula4** [reply]
Thank you!

And as for Snibgo's solution, is there a way to get that output converted tom PNG to TIF without the background showing again? I like having both solutions working as I have more data I will be looking at next week, and realistically I can't expect to come here for every step I get stuck at. You all are a huge help.

His was : 

magick weather.png ( +clone -colorspace HCL -channel 1 -separate
 +channel -threshold 50% ) -alpha off -compose CopyOpacity -composite out.png

I can't simply change that one to TIF at the end of the script, or I get the background again. I don't mind running a second command, I just don't know why the background won't stay gone.

Edit : I don't know how to explain this, but I changed out.png to out.bmp. In a separate command, I converted to TIF and the background is back again! I do not understand how.

**Reply by fmw42** [reply]
> I can't simply change that one to TIF at the end of the script, or I get the background again. I don't mind running a second command, I just don't know why the background won't stay gone.

Why not. It should work just fine.

```
magick weather.png ( +clone -colorspace HCL -channel 1 -separate
+channel -threshold 50% ) -alpha off -compose CopyOpacity -composite out.tif
```

**Reply by fmw42** [reply]
If you are running this in a bat script, then you have to double the % to %%

**Reply by formula4** [reply]
Yes, I have done that. I probably sound crazy, but let me show you :  
![image](https://user-images.githubusercontent.com/56749434/146661165-5c7ade70-a4a7-49fe-b247-54c8b02ba76f.png)

Step 1 : In that image, you can see my script open. I ran it on weather2.jpg. (You can see the file there)
Step 2 : You can see it wrote output.tif - and in the preview the file is transparent with the data.
Step 2 : I opened output.tif in photoshop, and the background is back!

What is going on here?

**_EDIT : I am starting to think Photoshop is bugging out with this. That is what it seems to be. To windows (using MSPAINT) the image is as expected. Windows Photo Viewer also shows it as expected. I will have to do some research on this. Today has been a bunch of weird things strung together. I will post here again if there is another issue. Thank you again!_**

**Reply by fmw42** [reply]
post your input image and your exact code

**Reply by fmw42** [reply]
Transparency may have to be specified differently for Photoshop. Tiff allows 3 different types of transparency. ImageMagick can set any of them using 

`-define tiff:alpha=associated|unassociated|unspecified`

**Comment by fmw42** [comment]
Both these work for me in my Mac in Unix syntax. They both open fine in Photoshop

```
magick 146649093-649ecf63-cb9b-432b-b969-ca43d8e52ed6.png -alpha off \
\( -clone 0 -color-threshold "srgb(20,175,0)-srgb(255,255,95)" \) \
\( -clone 0 -color-threshold "srgb(200,0,0)-srgb(255,200,255)" \) \
\( -clone 1,2 -compose plus -composite \) \
-delete 1,2 \
-alpha off -compose copy_opacity -composite \
-define tiff:alpha=associated result.tiff
```

<br>

```
magick 146649093-649ecf63-cb9b-432b-b969-ca43d8e52ed6.png -alpha off \
\( -clone 0 -color-threshold "srgb(20,175,0)-srgb(255,255,95)" \) \
\( -clone 0 -color-threshold "srgb(200,0,0)-srgb(255,200,255)" \) \
\( -clone 1,2 -compose plus -composite \) \
-delete 1,2 \
-alpha off -compose copy_opacity -composite \
-define tiff:alpha=unassociated result2.tiff

```
